addons/base_action_rule/models/base_action_rule.py

```diff
@@ -0,0 +1,386 @@
+    server_action_ids = fields.Many2many('ir.actions.server', string='Server Actions', domain="[('model_id', '=', model_id)]",
+                                         help="Examples: email reminders, call object service, etc.")
+    filter_pre_id = fields.Many2one("ir.filters", string='Before Update Filter', ondelete='restrict', domain="[('model_id', '=', model_id.model)]",
+                                    help="If present, this condition must be satisfied before the update of the record.")
+    filter_pre_domain = fields.Char(string='Before Update Domain',
+                                    help="If present, this condition must be satisfied before the update of the record.")
+    filter_id = fields.Many2one("ir.filters", string='Filter', ondelete='restrict', domain="[('model_id', '=', model_id.model)]",
+                                help="If present, this condition must be satisfied before executing the action rule.")
+    filter_domain = fields.Char(string='Domain', help="If present, this condition must be satisfied before executing the action rule.")
+    last_run = fields.Datetime(readonly=True, copy=False)
+    on_change_fields = fields.Char(string="On Change Fields Trigger", help="Comma-separated list of field names that triggers the onchange.")
+
+    # which fields have an impact on the registry
+    CRITICAL_FIELDS = ['model_id', 'active', 'kind', 'on_change_fields']
+
+    @api.onchange('model_id')
+    def onchange_model_id(self):
+        self.filter_pre_id = self.filter_id = False
+
+    @api.onchange('kind')
+    def onchange_kind(self):
+        if self.kind in ['on_create', 'on_create_or_write', 'on_unlink']:
+            self.filter_pre_id = self.trg_date_id = self.trg_date_range = self.trg_date_range_type = False
+        elif self.kind in ['on_write', 'on_create_or_write']:
+            self.trg_date_id = self.trg_date_range = self.trg_date_range_type = False
+        elif self.kind == 'on_time':
+            self.filter_pre_id = False
+
+    @api.onchange('filter_pre_id')
+    def onchange_filter_pre_id(self):
+        self.filter_pre_domain = self.filter_pre_id.domain
+
+    @api.onchange('filter_id')
+    def onchange_filter_id(self):
+        self.filter_domain = self.filter_id.domain
+
+    @api.model
+    def create(self, vals):
+        base_action_rule = super(BaseActionRule, self).create(vals)
+        self._update_cron()
+        self._update_registry()
+        return base_action_rule
+
+    @api.multi
+    def write(self, vals):
+        res = super(BaseActionRule, self).write(vals)
+        if set(vals) & set(self.CRITICAL_FIELDS):
+            self._update_cron()
+            self._update_registry()
+        return res
+
+    @api.multi
+    def unlink(self):
+        res = super(BaseActionRule, self).unlink()
+        self._update_cron()
+        self._update_registry()
+        return res
+
+    @api.multi
+    def _filter(self, action_filter, record_ids, domain=False):
```

Comment:
Can't we implement this method works on batch and return dictionary, checkout all calls of _filter, all are inside loop, so why not make this method handling multiple records.
@hmo-odoo what do you think as it may go in refactoring, should we consider here ?